src/jdk.jartool/share/classes/sun/tools/jar/Validator.java

```diff
@@ -62,20 +82,139 @@ final class Validator {
     private Set<String> concealedPkgs = Collections.emptySet();
     private ModuleDescriptor md;
     private String mdName;
+    private final ZipInputStream zis;
 
-    private Validator(Main main, ZipFile zf) {
+    private Validator(Main main, ZipFile zf, ZipInputStream zis) {
         this.main = main;
         this.zf = zf;
+        this.zis = zis;
         checkModuleDescriptor(MODULE_INFO);
     }
 
-    static boolean validate(Main main, ZipFile zf) throws IOException {
-        return new Validator(main, zf).validate();
+    static boolean validate(Main main, ZipFile zf, ZipInputStream zis) throws IOException {
+        return new Validator(main, zf, zis).validate();
+    }
+
+    /**
+     * Validate that the CEN/LOC file name header field adheres  to
+     * PKWARE APPNOTE-6.3.3.TXT:
+     *
+     * 4.4.17.1 The name of the file, with optional relative path.
+     * The path stored MUST not contain a drive or
+     * device letter, or a leading slash.  All slashes
+     * MUST be forward slashes '/' as opposed to
+     * backwards slashes '\' for compatibility with Amiga
+     * and UNIX file systems etc.
+     * Also validate that the file name is not "." and that any name element is
+     * not equal to ".."
+     *
+     * @param entryName CEN/LOC header file name field entry
+     * @return true if a valid Zip Entry file name; false otherwise
+     */
+    public static boolean isZipEntryNameValid(String entryName) {
+        return !INVALID_ZIP_ENTRY_NAME_PATTERN.matcher(entryName).find();
+    }
+
+    /**
+     * Helper class to deduplicate entry names.
+     * Keep a coutner for duplicate entry names and check if the entry name is valid on first add.
+     */
+    private class DedupEntryNames {
+        LinkedHashMap<String, Integer> entries = new LinkedHashMap<>();
+        boolean isCEN;
+
+        public DedupEntryNames(boolean isCEN) {
+            this.isCEN = isCEN;
+        }
+
+        public void add(String entryName) {
+            int latestCount = entries.merge(entryName, 1, (count, _) -> count + 1);
+            if (latestCount > 1) {
+                isValid = false;
+            } else if (!isZipEntryNameValid(entryName)) {
+                warn(formatMsg("error.validator.bad.entry.name", entryName));
+                isValid = false;
+            }
+        }
+
+        private boolean checkDuplicates(Map.Entry<String, Integer> counter) {
+            var count = counter.getValue();
+            if (count == 1) {
+            } else {
+                var msg = isCEN ? "warn.validator.duplicate.cen.entry" : "warn.validator.duplicate.loc.entry";
+                warn(formatMsg(msg, count.toString(), counter.getKey()));
+            }
+            return true;
```

Comment:
This seems to always be true as `warn()` results in `err.println` call (unless I am missing something)

Reply:
I should clean that up better. As we just want to show the message but not really filter out that entry.